ENIGMAsystem/SHELL/Platforms/General/UNIXfilemanip.cpp

```diff
@@ -47,7 +48,16 @@ int file_rename(string oldname,string newname)
 
 int file_copy(string fname,string newname)
 {
-  return system(("cp "+fname+" "+newname).c_str()); // Hackish, but there's no good implementation on Linux
+    ifstream srce(fname.c_str(),ios::binary);
```

Comment:
I hate to heckle you what you decided to keep, but this breaks link handling—`copy.c` is actually an incredibly complicated source file. You can take a look at it [in the coreutils source repository](http://git.savannah.gnu.org/cgit/coreutils.git/tree/src/copy.c#n1844). Look at that monster. Some thousand lines go into determining how to perform the copy operation we're requesting.

That said, the old function is a big security hole; I'm not sure if I wrote it or just approved it, but there are better ways to invoke `cp` than using the `system()` routine. You could do a fork-and-exec pattern to invoke it, like this:

```C++
pid_t c = fork();
if (c == -1) return false;
if (c) {
  int status;
  waitpid(c, &status, 0);
  return !status;
} else {
  execlp("cp", fname.c_str(), newname.c_str());
  exit(1);
}
```

Care to try that out?